### ISSUE TITLE
Rule proposal: no-non-null-property-assertion

### ISSUE THREAD

**Issue by studioromeo** [typescript-eslint/typescript-eslint#4247]
### Description

In typescript when `strictPropertyInitialization` is set class properties must be initialised with a value. 
Theres a few ways to handle this:
- declaring the property as possibly undefined
- giving it a default value 
- setting in the constructor
- Using the definite assignment operator ! to tell typescript we're going to assign it.

(see https://mariusschulz.com/blog/strict-property-initialization-in-typescript for more in-depth explanation).

The ! operator on class properties what I'd like the rule to target as this puts the onus on the developer to ensure that the property is assigned a value and cancels out some of the benefits of strict mode.

I thought the `no-non-null-assertion` rule would include class property definitions but it doesn't seem to fail on my version (4.33.0)

### Fail

```ts
class Foo {
    bar!: string;
}
```

### Pass

```ts
class Foo {
    bar?: string;
}
```

**Comment by bradzacher** [comment]
See this FAQ article: https://typescript-eslint.io/docs/linting/troubleshooting#how-can-i-ban-specific-language-feature

We don't want to build rules about banning specific language features unless it is an exceedingly common usecase or if there is some way we can add additional benefits.

In this instance - it's a rare thing for people to do, and the rule wouldn't have any additional benefits beyond just banning the syntax.
